sirepo/auth/ldap.py

```diff
@@ -0,0 +1,44 @@
+"""LDAP login
+:copyright: Copyright (c) 2018-2022 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+from pykern.pkdebug import pkdlog, pkdp
+from pykern.pkcollections import PKDict
+import ldap
+import sirepo.quest
+
+AUTH_METHOD = "ldap"
+
+#: User can see it
+AUTH_METHOD_VISIBLE = True
+
+
+class API(sirepo.quest.API):
+    @sirepo.quest.Spec("require_cookie_sentinel")
+    def api_authLdapAuthorized(self):
+        """Python-LDAP client authentication
+        Bind to server and authenticate
+        """
+        # Test User Credentials:
+        # username: Archimedes of Syracuse
+        # password: eureka
+
+        LDAP_SERVER = "ldap://10.10.10.10:389"
+        BASE_DN = "dc=example,dc=com"
+        OBJECT_TO_SEARCH = "objectClass=inetOrgPerson"
+        ATTRIBUTES_TO_SEARCH = ["uid"]
+
+        req = self.parse_post(type=False)
+
+        # initialize our connect to ldap-server
+        connect = ldap.initialize(LDAP_SERVER)
+
+        # bind to server and login
+        connect.simple_bind_s(
+            self._get_login(req.req_data.username), req.req_data.password
+        )
+
+        return PKDict()
+
+    def _get_login(self, username):
+        return "cn=" + username + ",ou=users,dc=example,dc=com"
```

Comment:
Syntax here should be guided by AD. It's likely it will be an email.

Reply:
> Are there injection attacks that can happen with this?

No. This is an argument to something that validates it. However, we wouldn't want to accept a 1GB user name or something so we should protect the input as with all inputs. We expect it to be characterized a particular way (e.g. less than 256 chars).

Reply:
> Syntax here should be guided by AD. It's likely it will be an email.

@robnagler what do you mean by AD? The syntax now assumes the login dn has an email.

Reply:
AD = Active Directory

> The syntax now assumes the login dn has an email.

I don't see this code. In any event, LDAP requires [escaping special characters](https://www.ibm.com/docs/en/sva/9.0.2.1?topic=tasks-valid-characters-ldap-user-group-names).

Reply:
Added front end warning when entering email/pass over 256 char, turned _get_login into _get_escaped and tested

Reply:
All web APIs should protect themselves. It's nice that the GUI let's the user know of the error, but the real problem is if someone hacks the API and submits invalid text.